File: solutions/Leetcode_1690/leetcode_1690.py
from itertools import accumulate
from typing import List


class Solution1:
    def stone_game_vii(self, stones: List[int]) -> int:
        n = len(stones)
        dp = [[0] * n for _ in range(n)]
        prev_sum = [0] + list(accumulate(stones))

        def dfs(i, j):
            if i == j:
                return 0
            if dp[i][j] == 0:
                sum = prev_sum[j+1] - prev_sum[i]
                dp[i][j] = max(sum - stones[i] - dfs(i+1, j), sum - stones[j] - dfs(i, j-1))
            return dp[i][j]
        return dfs(0, n-1)

File: solutions/Leetcode_1690/leetcode_1690.cc
#include <vector>

using namespace std;

class Solution1
{
public:
    int stoneGameVII(vector<int> &stones)
    {
        int n = stones.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));
        vector<int> prev_sum{0};
        for (int x : stones)
            prev_sum.push_back(prev_sum.back() + x);
        return dfs(0, n - 1, dp, prev_sum, stones);
    }

private:
    int dfs(int i, int j, vector<vector<int>> &dp, vector<int> &prev_sum, vector<int> &stones)
    {
        if (i == j)
            return 0;
        if (dp[i][j] == 0)
        {
            int summation = prev_sum[j + 1] - prev_sum[i];
            dp[i][j] = max(summation - stones[i] - dfs(i + 1, j, dp, prev_sum, stones), summation - stones[j] - dfs(i, j - 1, dp, prev_sum, stones));
        }
        return dp[i][j];
    }
};